src/inversion_scripts/lognormal_invert.py

```diff
@@ -0,0 +1,295 @@
+# Description: Script to perform inversion using lognormal errors
+# Usage: python lognormal_invert.py <path_to_config_file> <path_to_state_vector_file>
+
+# TODO: merge this script with invert.py to avoid redundancy
+# This script performs the inversion but using lognormal
+# errors instead of normal errors. As an alternative to
+# the invert.py script.
+
+import sys
+import yaml
+from itertools import product
+import numpy as np
+import xarray as xr
+from netCDF4 import Dataset
+from scipy.sparse import spdiags
+
+
+def lognormal_invert(config, state_vector_filepath, jacobian_sf):
+    """
+    Description:
+        Run inversion using lognormal errors following method from eqn 2 of
+        Chen et al., 2022 https://doi.org/10.5194/acp-22-10809-2022
+        Outputs inversion results to netcdf files.
+    Arguments:
+        state_vector_filepath [String] : path to state vector netcdf file
+        config                [Dict]   : dictionary of config variables
+    """
+    state_vector = xr.load_dataset(state_vector_filepath)
+    state_vector_labels = state_vector["StateVector"]
+    lats, lons = state_vector_labels.lat, state_vector_labels.lon
+
+    # used to determine convergence of xn 5e-3 is .5%
+    convergence_threshold = 5e-3
+
+    # Load in the observation and background data
+    ds = np.load("obs_ch4_tropomi.npz")
+    y = np.array(ds["obs_tropomi"])
+    ds = np.load("gc_ch4_bkgd.npz")
+    ybkg = np.array(ds["gc_ch4_bkgd"])
+
+    # We only solve using lognormal errors for state vector elements
+    # within the domain of interest, not the buffer elements, or the
+    # BC elements. So, to do this we split K into two matrices, one
+    # for the lognormal elements, and one for the normal elements.
+    optimize_bcs = config["OptimizeBCs"]
+    num_sv_elems = (
+        int(state_vector_labels.max().item()) + 4
+        if optimize_bcs
+        else state_vector_labels.max().item()
+    )
+    num_buffer_elems = int(config["nBufferClusters"])
+    num_normal_elems = num_buffer_elems + 4 if optimize_bcs else num_buffer_elems
+    ds = np.load("full_jacobian_K.npz")
+    K_temp = np.array(ds["K"]) * 1e9
+
+    # Apply scaling matrix if using precomputed Jacobian
+    if jacobian_sf is not None:
+        scale_factors = np.load(jacobian_sf)
+        # apply unit scaling for BC elements if optimizing BCs
```

Comment:
Is it possible for `jacobian_sf` to include elements for the BCs if the BCs were optimized in the previous inversion? Would the code here then append an additional (unneeded) four elements?

Reply:
This won't happen because `jacobian_sf` never has elements saved to it for the BCs when it is archived. This is because the way that we update the BCs in the posterior simulation is by updating `CH4_boundary_condition_ppb_increase_NSEW` directly in the HEMCO config file, rather than passing a file with the corrections in it like for emission and buffer elements. 

This does bring up a good point though, currently the BC optimization implementation resets the BC elements after each iteration (using the KF mode), and we are not iteratively passing in `CH4_boundary_condition_ppb_increase_NSEW` for successive prior simulations. Do you think we should be?

Reply:
Let's keep the current reset approach for both buffer emission elements and NSEW BC elements of the state vector (as discussed earlier).